rules/licenses_core.bzl

```diff
@@ -0,0 +1,187 @@
+            # it's on all targets.  Even some regular targets may be synthetic and
+            # not have the aspect. This provides protection against those outlier
+            # cases.
+            if provider in dep:
+                info = dep[provider]
+                if info.licenses:
+                    trans_licenses.append(info.licenses)
+                if info.deps:
+                    trans_deps.append(info.deps)
+                if info.traces:
+                    for trace in info.traces:
+                        traces.append("(" + ", ".join([str(ctx.label), ctx.rule.kind, name]) + ") -> " + trace)
+
+def gather_licenses_info_common(target, ctx, provider_factory, namespaces, filter_func):
+    """Collect license info from myself and my deps.
+
+    Any single target might directly depend on a license, or depend on
+    something that transitively depends on a license, or neither.
+    This aspect bundles all those into a single provider. At each level, we add
+    in new direct license deps found and forward up the transitive information
+    collected so far.
+
+    This is a common abstraction for crawling the dependency graph. It is parameterized
+    to allow specifying the provider that is populated with results. It is
+    configurable to select only licenses matching a certain namespace. It is also
+    configurable to specify which dependency edges should not be traced for the
+    purpose of tracing the graph.
+
+    Args:
+      target: The target of the aspect.
+      ctx: The aspect evaluation context.
+      provider_factory: abstracts the provider returned by this aspect
+      namespaces: a list of namespaces licenses must match to be included
+      filter_func: a function that returns true iff the dep edge should be ignored
+
+    Returns:
+      provider of parameterized type
+    """
+
+    # First we gather my direct license attachments
+    licenses = []
+    if ctx.rule.kind == "_license":
+        # Don't try to gather licenses from the license rule itself. We'll just
+        # blunder into the text file of the license and pick up the default
+        # attribute of the package, which we don't want.
+        pass
+    else:
+        if hasattr(ctx.rule.attr, "applicable_licenses"):
+            for dep in ctx.rule.attr.applicable_licenses:
+                if LicenseInfo in dep:
+                    lic = dep[LicenseInfo]
+
+                    # This check shouldn't be necessary since any license created
+                    # by the official code will have this set. However, one of the
+                    # tests has its own implementation of license that had to be fixed
+                    # so this is just a conservative safety check.
+                    if hasattr(lic, "namespace"):
+                        if lic.namespace in namespaces:
+                            licenses.append(lic)
+                    else:
+                        fail("should have a namespace")
```

Comment:
This logic is enforcing that the `namespace` attr of the `license` rule is required. Should this logic instead be implemented in the `license` rule definition with: `mandatory = True`?